projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+)
+
+// Comments ----------
+// I write a lot of comments, it has helped me the entire course to learn, understand and remember
+// They can all be removed at the end to clean it up
+
+// Error Messages ----------
+// The Errors are deliberately numbered [x] so I could follow them up the call stack
+// They can be replaced with "user friendly" messages later
+
+func parseResponseBody(response *http.Response) (string, error) {
+	// try to read the response body
+	responseBody, err := io.ReadAll(response.Body)
+
+	// handle any error immediately
+	if err != nil {
+		return "", fmt.Errorf("[2] failed to read response body : %w", err)
+	}
+
+	// use the defer keyword early to close the body regardless
+	defer response.Body.Close()
+
+	return string(responseBody), nil
+}
+
+// returned integer is the retry duration in seconds
+func parseRetryAfterHeader(retryAfterHeaderString string) (int, error) {
+	// first condition "a while"
+	if retryAfterHeaderString == "a while" {
+		return 0, fmt.Errorf("[3] retry-after header is 'a while'")
+	}
+
+	// second condition an integer in string format
+	retryAfterInteger, err := strconv.Atoi(retryAfterHeaderString)
+
+	// should immediately error check here...
+	// but i need to be able to continue after...
+	// in order to check if it's a timestamp...?
+	// i am trying to avoid nested if logic
+
+	// if err != nil {
+	//  return 0, nil
+	// }
+
+	// if we successfully parse an integer return it
+	if err == nil {
+		return retryAfterInteger, nil
+	}
+
+	// third condition a timestamp in string format
+	retryAfterTimeStamp, err := time.Parse(time.RFC1123, retryAfterHeaderString)
```

Comment:
can you look into `time.RFC1123` vs `http.TimeFormat`? are they the same?

Reply:
I looked into it... `http.TimeFormat` deals with more than just RFC1123 there are apparently multiple RFC specs that HTTP can use in the headers... so using `http.TimeFormat` makes it more robust